mscore/plugin/api/selection.cpp

```diff
@@ -28,5 +29,45 @@ Selection* selectionWrap(Ms::Selection* select)
       return w;
       }
 
+//---------------------------------------------------------
+//   QmlSelElementsListAccess::at
+//---------------------------------------------------------
+
+Element* QmlSelElementsListAccess::at(QQmlListProperty<Element>* l, int i)
+      {
+      Ms::Selection* sel = static_cast<Ms::Selection*>(l->data);
+      return wrap(sel->elements()[i]);
+      }
+
+//---------------------------------------------------------
+//   QmlSelElementsListAccess::count
+//---------------------------------------------------------
+
+int QmlSelElementsListAccess::count(QQmlListProperty<Element>* l)
+      {
+      Ms::Selection* sel = static_cast<Ms::Selection*>(l->data);
+      return sel->elements().length();
+      }
+
+//---------------------------------------------------------
+//   QmlSelElementsListAccess::clear
+//---------------------------------------------------------
+
+void QmlSelElementsListAccess::clear(QQmlListProperty<Element>* l)
+      {
+      Ms::Selection* sel = static_cast<Ms::Selection*>(l->data);
+      sel->clear();
+      }
+
+//---------------------------------------------------------
+//   QmlSelElementsListAccess::append
+//---------------------------------------------------------
+
+void QmlSelElementsListAccess::append(QQmlListProperty<Element>* l, Element *e)
```

Comment:
If you have this append function, what about the remove function (https://github.com/musescore/MuseScore/blob/f495ca33b2274db471946f92e660e27122e0fe3f/libmscore/select.h#L184).
I believe you would only have to change 'sel->add(e->element());' for 'sel->remove(e->element());'

Reply:
The above four methods are the only ones supported for QML's standard Javascript array exposure. Under the associated MuseScore issue there should be some test scripts that illustrate how to clear and change arrays of selections using javascript.

-Dale

Reply:
too bad...

Reply:
We can implement Q_INVOKABLE functions in `Selection` class itself instead so we won't be limited to operations that are supported by QQmlListProperty. Having explicit functions for selection operations may also be preferable since it should probably be more clear and easy to use (something like `sel.add(e)` as opposed to `sel.elements.push(e)`), and allows us also to add operations like `selectRange()` or `selectSingle()` (i.e. deselect all and select the given element).

Reply:
So, if i understand well, the only thing needed would be a Q_INVOKABLE in front of every fonction needed.
https://github.com/musescore/MuseScore/blob/f495ca33b2274db471946f92e660e27122e0fe3f/libmscore/select.h#L184
would become 'Q_INVOKABLE void remove(Element*);'
etc.
If adding functions to the API is as simple as adding a word in front of each of them and documenting it, how come no one knew it already! There would be tons of pull requests!

Reply:
Unfortunately that is not that simple :) Functionality exposed to plugins has some specific set of requirements (all changes to score data must be undoable, exposed function must be better protected against incorrect usage to prevent crashes and data loss where possible, they should correctly handle objects' ownership (QML vs. C++) to prevent massive memory leaks etc.) and should generally have a more stable and easier to use API than we are able to afford in libmscore itself at the moment (I have also described something similar [here](https://musescore.org/en/node/293837#comment-942176) in point 3 of my comment).

In an attempt to make these issues easier to resolve in the long run, plugins API was decoupled from libmscore (its implementation is mainly located [here](https://github.com/musescore/MuseScore/tree/master/mscore/plugin/api) now), and now QML plugins code deals with wrappers which provide specific API rather than directly with the real objects from libmscore. For that reason appending `Q_INVOKABLE` in front of libmscore functions won't do any desired effect as those functions simply operate on different kind of objects. But even if it would work, it would in fact be incorrect to do so since these functions don't fully satisfy the requirements for plugin-exposed functionality. In case of deselection what we are more likely to need is something like [this code](https://github.com/musescore/MuseScore/blob/f495ca33b2274db471946f92e660e27122e0fe3f/libmscore/score.cpp#L2934) which would also take care of setting up all the relevant flags to notify the attached views and any other components of changes that were made to the selection. Of course, libmscore often lacks consistency here and part of the listed issues can be resolved by restructuring libmscore itself (which would take a relatively large effort currently) but, in general, we still need a separate implementation for the functionality exposed to plugins, and directly exposing libmscore interface won't work in many cases.

But for implementing an API for plugins that satisfies those requirements, pull requests are certainly welcome!